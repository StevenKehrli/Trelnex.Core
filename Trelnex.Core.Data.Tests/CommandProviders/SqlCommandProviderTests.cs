using Azure.Core;
using Azure.Identity;
using Microsoft.Data.SqlClient;
using Microsoft.Extensions.Configuration;

namespace Trelnex.Core.Data.Tests.CommandProviders;

[Ignore("Requires a SQL server.")]
public class SqlCommandProviderTests : CommandProviderTests
{
    private readonly string _scope = "https://database.windows.net/.default";

    private TokenCredential _tokenCredential = null!;
    private string _connectionString = null!;
    private string _tableName = null!;

    [OneTimeSetUp]
    public async Task TestFixtureSetup()
    {
        // This method is called once prior to executing any of the tests in the fixture.

        // create the test configuration
        var configuration = new ConfigurationBuilder()
            .AddJsonFile("appsettings.json", optional: true, reloadOnChange: true)
            .AddJsonFile("appsettings.User.json", optional: true, reloadOnChange: true)
            .Build();

        var providerConfiguration = configuration.GetSection("SqlCommandProviders").Get<SqlCommandProviderConfiguration>()!;

        var scsBuilder = new SqlConnectionStringBuilder()
        {
            DataSource = providerConfiguration.DataSource,
            InitialCatalog = providerConfiguration.InitialCatalog,
            Encrypt = true,
        };

        _connectionString = scsBuilder.ConnectionString;
        _tableName = providerConfiguration.TableName;

        // create the command provider
        _tokenCredential = new DefaultAzureCredential();

        var sqlClientOptions = new SqlClientOptions(
            TokenCredential: _tokenCredential,
            Scope: _scope,
            DataSource: providerConfiguration.DataSource,
            InitialCatalog: providerConfiguration.InitialCatalog,
            TableNames: [ providerConfiguration.TableName ]
        );

        var factory = await SqlCommandProviderFactory.Create(
            sqlClientOptions);

        _commandProvider = factory.Create<ITestItem, TestItem>(
            providerConfiguration.TableName,
            "test-item",
            TestItem.Validator,
            CommandOperations.All);
    }

    [TearDown]
    public void TestCleanup()
    {
        // This method is called after each test has run.
        using var sqlConnection = new SqlConnection(_connectionString);

        var tokenRequestContext = new TokenRequestContext([ _scope ]);
        sqlConnection.AccessToken = _tokenCredential.GetToken(tokenRequestContext, default).Token;

        sqlConnection.Open();

        var cmdText = $"DELETE FROM [{_tableName}-events]; DELETE FROM [{_tableName}];";
        var sqlCommand = new SqlCommand(cmdText, sqlConnection);

        sqlCommand.ExecuteNonQuery();
    }

    /// <summary>
    /// Represents the configuration properties for SQL command providers.
    /// </summary>
    private record SqlCommandProviderConfiguration(
        string DataSource,
        string InitialCatalog,
        string TableName);
}
