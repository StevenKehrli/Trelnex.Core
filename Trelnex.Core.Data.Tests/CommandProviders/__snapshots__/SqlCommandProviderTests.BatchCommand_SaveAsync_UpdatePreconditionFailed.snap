{
  "saved": [
    {
      "HttpStatusCode": "FailedDependency",
      "ReadResult": null
    },
    {
      "HttpStatusCode": "InternalServerError",
      "ReadResult": null
    }
  ],
  "read": [
    {
      "Item": {
        "Id": "3107d2ff-9f33-4718-b180-f7b65223fba8",
        "PartitionKey": "90eabddb-a71a-4ccb-aa1a-062590597fb2",
        "TypeName": "test-item",
        "CreatedDate": "2025-01-10T17:09:50.981035Z",
        "UpdatedDate": "2025-01-10T17:09:50.981035Z",
        "DeletedDate": null,
        "IsDeleted": null,
        "ETag": "D8881DEB-91E2-4F42-B33D-61E71AAF852E",
        "Message": "Message #1"
      }
    },
    {
      "Item": {
        "Id": "5f969f1c-c115-4982-949b-4e4987197dfe",
        "PartitionKey": "90eabddb-a71a-4ccb-aa1a-062590597fb2",
        "TypeName": "test-item",
        "CreatedDate": "2025-01-10T17:09:50.981105Z",
        "UpdatedDate": "2025-01-10T17:09:51.971816Z",
        "DeletedDate": null,
        "IsDeleted": null,
        "ETag": "4A7921B0-A051-434D-B853-90E33CB61F5D",
        "Message": "Message #0"
      }
    }
  ]
}
